mewarpx/mewarpx/mepicmi.py

```diff
@@ -53,3 +54,65 @@ def init(self, kw):
                 grid=self.grid
             )
         )
+
+    def init_pid_dict(self):
+        """Function to build a dictionary of all the extra particle attributes
+        (and weight). This has to happen after warpx has been initialized
+        so that the particle container already exists.
+        """
+        self.pid_dict = {'w':0}
+        self.nattribs = _libwarpx.get_nattr()
+        # npids are used to inform the dimension of the attr array (used in
+        # add_particles), so we need to subtract the non-extra pid arrays
+        # (which include the weight) and then add 1 since the weight is treated
+        # as an extra pid by AddNParticles()
+        self.npids = _libwarpx.get_nattr_species(self.name) - self.nattribs + 1
+
+    def add_pid(self, pid_name):
+        """Wrapper to add a new PID to the particle data arrays at runtime.
+
+        Arguments:
+            pid_name (str): Name of the new PID.
+        """
+        # check if PID already exists
+        if pid_name in self.pid_dict:
+            return
```

Comment:
Would it be useful to add a message saying that the PID already exists?

Reply:
I can go either way here. I can imagine there would be cases where the same PID should be added from different places if the same species is involved. This is fine as long as the PID isn't actually repeatedly added which this check prevents. I don't really think we need a warning since it isn't problematic.

Reply:
Ah I see, in that case we don't need it then.